### PR TITLE
Lifting variable on retrace

### LINKED ISSUE #27120: tf.function-decorated function tried to create variables on non-first call
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Debian Testing
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below): 2.0.0.dev20190227
- Python version: 3.7

**Describe the current behavior**

A function which correctly works when in eager mode does not work anymore when annotated with `tf.function`.

In particular, it complains about `ValueError: tf.function-decorated function tried to create variables on non-first call.`, even though the function is always called with different parameters.

This is a continuation of https://github.com/tensorflow/tensorflow/issues/26812#issuecomment-475600836.

**Describe the expected behavior**

The `apply_gradients_once()` function should work even when annotated with `tf.function`.

**Code to reproduce the issue**

```python3
import tensorflow as tf
import numpy as np


fast_optimizer = tf.keras.optimizers.Adam(
        learning_rate=1e-3)

slow_optimizer = tf.keras.optimizers.Adam(
        learning_rate=1e-3 * 1e-9)


@tf.function
def apply_gradients_once(optimizer, grads, vars):
    grads = [grads]
    optimizer.apply_gradients(zip(grads, vars))


def apply_grads(use_fast, grads_per_model, vars):
    for i in range(2):
        if use_fast[i]:
            apply_gradients_once(fast_optimizer, grads_per_model[i], vars[i])
        else:
            apply_gradients_once(slow_optimizer, grads_per_model[i], vars[i])


def compute_loss(w, x, y):
    r = (w * x - y)**2
    r = tf.math.reduce_mean(r)
    return r

def compute_gradients(model):
    with tf.GradientTape() as tape:
        tape.watch(model)
        loss = compute_loss(model, x, y)
    grads = tape.gradient(loss, model)
    return grads


w = [
    tf.Variable(0.0),
    tf.Variable(1.0)]

x = np.array([1, 2, 3])
y = np.array([1, 2, 3])

vars = []
grads = []
for i in range(2):
    vars.append([w[i]])
    grads.append(compute_gradients(w[i]))

apply_grads([True, False], grads, vars)
```

**Other info / logs**

Error log:

```
Traceback (most recent call last):
  File "main.py", line 52, in <module>
    apply_grads([True, False], grads, vars)
  File "main.py", line 23, in apply_grads
    apply_gradients_once(slow_optimizer, grads_per_model[i], vars[i])
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/eager/def_function.py", line 414, in __call__
    return self._stateless_fn(*args, **kwds)  # pylint: disable=not-callable
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/eager/function.py", line 1254, in __call__
    graph_function, args, kwargs = self._maybe_define_function(args, kwargs)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/eager/function.py", line 1577, in _maybe_define_function
    args, kwargs, override_flat_arg_shapes=relaxed_arg_shapes)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/eager/function.py", line 1479, in _create_graph_function
    capture_by_value=self._capture_by_value),
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/func_graph.py", line 685, in func_graph_from_py_func
    func_outputs = python_func(*func_args, **func_kwargs)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/eager/def_function.py", line 317, in wrapped_fn
    return weak_wrapped_fn().__wrapped__(*args, **kwds)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/func_graph.py", line 677, in wrapper
    ), args, kwargs)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/autograph/impl/api.py", line 392, in converted_call
    result = converted_f(*effective_args, **kwargs)
  File "/tmp/tmpr2ti5o1e.py", line 4, in tf__apply_gradients_once
    ag__.converted_call('apply_gradients', optimizer, ag__.ConversionOptions(recursive=True, verbose=0, strip_decorators=(tf.function, defun, ag__.convert, ag__.do_not_convert, ag__.converted_call), force_conversion=False, optional_features=(), internal_convert_user_code=True), (zip(grads, vars),), {})
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/autograph/impl/api.py", line 267, in converted_call
    return _call_unconverted(f, args, kwargs)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/autograph/impl/api.py", line 188, in _call_unconverted
    return f(*args, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/optimizer_v2/optimizer_v2.py", line 399, in apply_gradients
    self._create_hypers()
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/optimizer_v2/optimizer_v2.py", line 558, in _create_hypers
    aggregation=tf_variables.VariableAggregation.ONLY_FIRST_REPLICA)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/optimizer_v2/optimizer_v2.py", line 727, in add_weight
    aggregation=aggregation)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/training/tracking/base.py", line 622, in _add_variable_with_custom_getter
    **kwargs_for_getter)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/engine/base_layer_utils.py", line 152, in make_variable
    aggregation=aggregation)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/ops/variables.py", line 212, in __call__
    return cls._variable_v1_call(*args, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/ops/variables.py", line 175, in _variable_v1_call
    aggregation=aggregation)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/ops/variables.py", line 58, in getter
    return captured_getter(captured_previous, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/eager/def_function.py", line 375, in invalid_creator_scope
    "tf.function-decorated function tried to create "
ValueError: tf.function-decorated function tried to create variables on non-first call.
```

### PR DESCRIPTION
**Pull request by bhack**
Explore the effect on tests to fix: https://github.com/tensorflow/tensorflow/issues/27120

### CODE REVIEW

**Comment by bhack** [comment]
@mdanatg As supposed we have 3 tests failing. 

https://github.com/tensorflow/tensorflow/blob/49e6961ee63622bd2b3d7b8654b06ac8c54efd17/tensorflow/python/eager/def_function_test.py#L76-L83

https://github.com/tensorflow/tensorflow/blob/49e6961ee63622bd2b3d7b8654b06ac8c54efd17/tensorflow/python/eager/def_function_test.py#L85-L94

https://github.com/tensorflow/tensorflow/blob/dee93aa2d97ac276099bfe414c2911f4be2dd3c2/tensorflow/python/ops/parallel_for/control_flow_ops_test.py#L2431-L2444

**Comment by bhack** [comment]
What we want to do here?

**Comment by mdanatg** [comment]
The purpose of the PR was just to verify if by any chance it has a simple fix, right? The test results indicate that a straightforward refactoring doesn't work, so not much left to do but close the PR. A larger refactoring of the function objects is needed instead, but that's a significant project. We've been discussing it internally as well, but haven't started doing it yet.

**Comment by bhack** [comment]
> The purpose of the PR was just to verify if by any chance it has a simple fix, right?

Partially yes but also to find any action item on how to refactoring this. But if it is too complex or you want to talk about this internally it is ok.

**Comment by mdanatg** [comment]
That we're discussing internally doesn't matter much. It's mainly the caution about complexity: in short, we have the polymorphic Function object doing the variable lifting, but it should be individual concrete functions who do it instead. And then we'd need to fix whatever issues come from refactoring that. Would definitely be useful to do, but not easy to pull through.

**Comment by mdanatg** [comment]
To be clear, if you're interested in giving it a shot, it would be welcome!

**Comment by bhack** [comment]
I was a little be confused by `_initialize`:

At the end of the function it seems that it is always going to raise the `ValueError`. Where is the alternative execution path of `_initialize`? ;

```python
    def invalid_creator_scope(*unused_args, **unused_kwds):
      """Disables variable creation."""
      raise ValueError(
          "tf.function-decorated function tried to create "
          "variables on non-first call.")

    self._stateless_fn = self._defun_with_scope(invalid_creator_scope)
```

**Comment by mdanatg** [comment]
It will always raise an error if the function creates any variable. This is just half of the implementation - there is another branch of the code (different class, IIRC?) which doesn't have this check.

Loosely speaking, there's the following logic:

* 1st trace: a function is created without and invalid_creator_scope; then a second trace is immediately triggered
* subsequent traces: another function is created, with invalid_creator_scope - this is the one we're looking at

It's been a while since I looked at the code and can't remember exactly what is where, but I'd start with tracking the `__call__` method of `Function` in `def_function.py` - that's the main entry point.

**Comment by bhack** [comment]
@mdanatg I've added your test and hacked a little bit the `_call`. 
This was just an hack to execute tests but it seems that the new test and all the other existing tests are passing.

Is there something that it is not covered with existing tests?

**Comment by mdanatg** [comment]
Hmm, if tests are passing, that's good news. I do suspect the coverage is not great, but it's still very encouraging. Let's try the following then: put all these changes behind a global flag (can just make a constant in def_function); set its value to False (disabled), and then let's merge the PR.
Once merged, I can flip the flag and run a much larger suite of tests that we can access internally.

**Comment by bhack** [comment]
> I do suspect the coverage is not great, 
Yes I had the same suspect, In any case I've submitted the constant flag.

File: tensorflow/python/eager/def_function.py
Comment (mdanatg) [inline tensorflow/python/eager/def_function.py:102]: Naming nit: let's be more specific with this one. Can add the PR/issue number as a comment. How about: `ALLOW_DYNAMIC_VARIABLE_CREATION = False`

File: tensorflow/python/eager/def_function_test.py
Comment (mdanatg) [inline tensorflow/python/eager/def_function_test.py:268]: Here, let's write two tests, one which overrides the flag to True, the other overriding it to False. That way the tests remain independent of any future changes, and we avoid the "if this test that", which is a bit of an antipattern.

**Review by mdanatg — COMMENTED** [review]
Thanks! Couple of loose ends and then should be good.

File: tensorflow/python/eager/def_function.py
Comment (bhack) [reply tensorflow/python/eager/def_function.py]: Yes I supposed that PR_49310 was for you short term test. Now is `ALLOW_DYNAMIC_VARIABLE_CREATION`

File: tensorflow/python/eager/def_function_test.py
Comment (bhack) [reply tensorflow/python/eager/def_function_test.py]: Now we have a parametrized test that it is failing in the FALSE case. Is this what you want or do you want that the test passes in both cases? In that case we need to totally separate this in two tests and in the false case expect `ValueError:` that is the current behavior in master with false.

Comment (mdanatg) [reply tensorflow/python/eager/def_function_test.py]: Yes, the test should expect a ValueError in the False case.

Comment (bhack) [reply tensorflow/python/eager/def_function_test.py]: OK, done

**Review by mdanatg — APPROVED** [review]
Thanks!

**Comment by bhack** [comment]
Just fixed a pylint long line

**Review by mdanatg — APPROVED** [review]

**Comment by bhack** [comment]
Ok only `import/copybara` is on hold... let me know when you have any news.

**Comment by bhack** [comment]
It Is closed but the default rlag is false.

**Comment by mdanatg** [comment]
@bhack  One potential problem I came across is that when the flag is on, function invocation always uses the the function's `_stateful_fn`, which is slower than the stateless counterpart because it uses conditionals for each variable that's being read.

I'm seeing some some training tests that use fixed random produce different results, but I'm not sure whether the test is just brittle and calling the other function is enough to break determinism or there's is something else going on.

It would be interesting to test with a small Keras Sequential model, and test (1) the performance and (2) the determinism of model.compile/model.train when given some fixed input data and a fixed random seed.

**Comment by bhack** [comment]
> It would be interesting to test with a small Keras Sequential model, and test (1) the performance and (2) the determinism of model.compile/model.train when given some fixed input data and a fixed random seed.

I suppose that we have many of these tests in the repo.

Where we could cherry pick one of these and add the function that you want to test?

**Comment by mdanatg** [comment]
I think a practical option could be to set the flag from an environment variable, wait until the change gets into tf-nightly, then run some of the models in https://github.com/tensorflow/models with that version, toggling the flag to True through the env var.

**Comment by bhack** [comment]
Mhh.. if you want to test perf reg on a e2e real model what is the status of the [internal benchmarks](https://github.com/tensorflow/build/issues/10)?

**Comment by bhack** [comment]
@mdanatg @sganeshb Can you check internally if we could use that benchmark?

**Comment by sganeshb** [comment]
We have some candidate scripts for a few use cases, that we occassionally share with external collaborators.
These scripts can run the tensorflow/models benchmarks on a GPU machine.

Example: https://github.com/tensorflow/benchmarks/blob/master/perfzero/dockertest/resnet50_synth.sh

Do you have a model/benchmark of interest? We can setup an example script for you to try.

**Comment by bhack** [comment]
@sganeshb Thank you. For this specific PR I think you can sync with @mdanatg internally.

More in general it could be nice to run a very small selection in the TF CI (/cc @mihaimaruseac) just as a canary to support PRs reviews (see https://github.com/tensorflow/build/issues/10).

**Comment by sganeshb** [comment]
Thanks for the feedback.
For the short term, feel free to re-purpose this https://github.com/tensorflow/benchmarks/blob/master/perfzero/dockertest/resnet50_synth.sh to test out your env var on resnet50.

We'll look into providing a general test added to TF CI. We'll need to discuss internally on priorities given internal staffing and other resources for these efforts.

**Comment by sumanthratna** [comment]
Hi @bhack and @sganeshb, can you explain why the following code raises `ValueError: tf.function-decorated function tried to create variables on non-first call`? 

```python
```

CC @sjtusmartboy

**Comment by mdanatg** [comment]
We don't, at least not for now. We're working to remove this limitation, but  we're forced to proceed slowly beacuse of the complexity of the issue, as well as the backward compatibility restrictions.

Only this syntax is expected to work:

```
vars = {}

@def_function.function
def myfunc(val, key):
    if key not in vars:
      vars[key] = variables.Variable(val)
```

I can see why the name of the constant might be confusing, though; perhaps we should rename it.

**Comment by bhack** [comment]
@mdanatg Do you will accept a failing test for this specific like case like in https://github.com/tensorflow/tensorflow/pull/51374 ?

**Comment by mdanatg** [comment]
Yep! I'm surprised we don't already have one.

**Comment by bhack** [comment]
@sumanthratna I've added your use case as an expected failing test at https://github.com/tensorflow/tensorflow/pull/51538
The reference bug is always https://github.com/tensorflow/tensorflow/issues/27120

**Pull request merged**
